vmm/src/vm.rs

```diff
@@ -435,19 +437,22 @@ impl DeviceManager {
                 .map_err(DeviceManagerError::RegisterIoevent)?;
         }
 
-        // Assign IRQ to the virtio-pci device
-        let irqfd = EventFd::new(EFD_NONBLOCK).map_err(DeviceManagerError::EventFd)?;
-        let irq_num = allocator
-            .allocate_irq()
-            .ok_or(DeviceManagerError::AllocateIrq)?;
-        vm_fd
-            .register_irqfd(irqfd.as_raw_fd(), irq_num)
-            .map_err(DeviceManagerError::Irq)?;
+        let vm_fd_clone = vm_fd.clone();
 
-        let irq_cb = Arc::new(Box::new(move || irqfd.write(1)) as IrqClosure);
+        let msi_cb = Arc::new(Box::new(move |entry: MsixTableEntry| {
+            let msi_queue = kvm_msi {
+                address_lo: entry.msg_addr_lo,
+                address_hi: entry.msg_addr_hi,
+                data: entry.msg_data,
+                flags: 0u32,
+                devid: 0u32,
+                pad: [0u8; 12],
+            };
+
+            vm_fd_clone.signal_msi(msi_queue)
```

Comment:
This ioctl has CAP dependency(KVM_CAP_SIGNAL_MSI) I guess we need to check that somewhere.

Reply:
Yes I'll add this check, thanks for reminding me :)